### ISSUE TITLE
Cannot create a project

### ISSUE THREAD

**Issue by Fancyflame** [dotnet/sdk#14867]
Problem encountered on https://dotnet.microsoft.com/learn/dotnet/hello-world-tutorial/install
Operating System: android(ubuntu environment)

When I tried to create a project this was occurred
![Screenshot_2020-12-07-23-49-00](https://user-images.githubusercontent.com/31250625/101372496-ef19a180-38e6-11eb-9222-172113be6704.png)

I have tried to running other softwares on my android device, Rust, Go and C++ work well on the ubuntu environment, so it might be not something wrong with the environment.
Besides, I think my storage is large enough, which has at least 10GB free.
This is the information

![Screenshot_2020-12-07-23-56-25](https://user-images.githubusercontent.com/31250625/101373442-e7a6c800-38e7-11eb-8e6b-dcf8a45b25a1.png)

So is there a way to resolve it? Thanks. :)

**Comment by mairaw** [comment]
Transferred the issue @timheuer

**Comment by Fancyflame** [comment]
> @leecow @KathleenDollard @timheuer have you seen this issue before? Also, have we tested this scenario before of running dotnet on Ubuntu on an Android device? I don't know if this is a supported scenario or not.

If you need to test how it works on Android, it is suggested that, just like me, install Termux, then install Ubuntu(https://github.com/Neo-Oli/termux-ubuntu) in it, after that install dotnet via the manual method.

**Comment by timheuer** [comment]
@Fancyflame in this configuration do you have space limitations in the Ubuntu instances installed?

**Comment by Fancyflame** [comment]
> @Fancyflame in this configuration do you have space limitations in the Ubuntu instances installed?

No. Even if there is, at least the space Ubuntu can use is much enough.

**Comment by joeloff** [comment]
I wonder if this is a memory issue, rather than disk space. Is it possible to run something like ```free -m```?

**Comment by Fancyflame** [comment]
@joeloff 
free -m
```
total        used        free      shared  buff/cache   available
Mem:           2791        1696          61          11        1032         801
Swap:           255         165          90
```

> Or maybe run something like `top` or pull usage data from /proc when running the `new` command to create a project and see if there's a spike in memory usage.
I'm sorry for be poor in English that I can't recognize what the above means, could you please explain it? 😂

**Comment by joeloff** [comment]
No, it shouldn't require that much memory. ```-m``` reports available memory in MB, so it looks like you have ~1.6GB available.

As for my other comment, I was just thinking there might be a way to monitor the processes while dotnet is running and trying to create the template.

Just running ```dotnet new``` should print a list of all the available templates. Does the same error occur if you try something like ```dotnet new console```?

At least on Windows, if you were running from a folder, e.g. C:\test, it uses the ```test``` folder name for the project. Though you did run it from "~".

Can you try creating a directory and run the ```dotnet new``` from inside the directory?

**Comment by Fancyflame** [comment]
@joeloff 
I created a new directory and run `dotnet new console ` in it the error still occurred.
Besides, my current SDK is .NET Core 3.1. I tried install .NET 5 and run it, which is even worse. It only print `Segmentation fault` and exit. Afterward as a regular user (I used to operate as root user), I ran .NET 5 in the same way , it magically no longer print `Segmentation fault` but still print the error of lacking in storage space.   😕???

**Comment by joeloff** [comment]
@stephentoub I was poking around in the runtime repo and noticed you made a number of commits around the mutex code and wondered whether you might have any insights here. When CreateMutexCore is called, are there any operations that can result in the IOException mentioned in this issue where it claims to be out of storage?

Based on the explicit exceptions it can throw, my guess is the exception is a result of this https://github.com/dotnet/runtime/blob/be27e294a12514638bb5337c7fe5cc72d93a443a/src/libraries/System.Private.CoreLib/src/System/Threading/Mutex.Windows.cs#L35
 and ultimately https://github.com/dotnet/runtime/blob/be27e294a12514638bb5337c7fe5cc72d93a443a/src/libraries/Common/src/System/IO/Win32Marshal.cs#L62

**Comment by kouvel** [comment]
The shared memory files go into `/tmp` in the non-Android Linux builds. Does `/tmp` exist, or is the tmp directory located at `/data/local/tmp`? I'm not sure if there is an official Android build for .NET Core. If that's the issue it might be possible to work around it by symlinking `/tmp` to `/data/local/tmp`.

**Comment by joeloff** [comment]
Thanks @kouvel and @stephentoub 

@Fancyflame please let us know if the suggestion worked. Your scenario isn't officially supported, but I'm hoping we can help

**Comment by Fancyflame** [comment]
@kouvel @joeloff  Where I run dotnet is a tiny Ubuntu virtual machine, there is `/tmp` directory already but even no `/data`. I think creating `/data/local/tmp` and linking `/tmp` to it doesn't make sense.

@stephentoub I'm not very familiar with Linux, so I dont know what to do according to your explaination. Could u please tell me the what to do In more detail?

**Comment by Fancyflame** [comment]
@joeloff Yes I know Android platform was not being supported, but the Ubuntu virtual machine is aimed to create a environment in which common Linux software could working, and as expected most softwares worked perfectly well, includes Vim, Python, Nodejs, Golang, Rust and C++. Undeniable due to the imperfection of the environment there are some softwares runned in to problems, but I still hope dotnet is not in the list. :)

**Comment by kouvel** [comment]
Oh ok, if `/tmp` exists and has the appropriate permissions that should be fine. Could you please run the following and share the log file?
```bash
strace -o trace.log -e \!read,lseek,mprotect dotnet new
```

**Comment by Fancyflame** [comment]
@kouvel Same as before, except one more log:
`proot warning: ptrace request 'PTRACE_???' not supported yet`

**Comment by kouvel** [comment]
Did it create a `trace.log` file? Wonder if that's just a warning or if `strace` is not working on that machine.

**Comment by Fancyflame** [comment]
@kouvel Do you want to see that? Its content is rather much! I think this problem may caused by proot's low version, but I cannot update proot due to the low Android version which is Android 6. Packages for this version are no longer maintained.

**Comment by kouvel** [comment]
The log should show the system calls at the point of failure and it might be useful, the proot issue may be separate and the relevant logs may still be there. If you can upload the log and send me a link I can take a look.

**Comment by Fancyflame** [comment]
@kouvel 
I'm so sorry for reply such lately. Here is the file. [trace.log](https://github.com/dotnet/sdk/files/5826342/trace.log)
In fact, because of Rust's outstanding compatibility and performance, I chose Rust instead of C# eventually, hence fixing this problem is not urgent for me now and that's why I forgot to reply you. Still, I'm willing to settle this problem.

**Comment by kouvel** [comment]
Thanks @Fancyflame. It looks an `flock` system call is failing with `EACCES`:

```
newfstatat(AT_FDCWD, "/tmp/", {st_mode=S_IFDIR|0700, st_size=4096, ...}, 0) = 0
newfstatat(AT_FDCWD, "/tmp/.dotnet", {st_mode=S_IFDIR|0777, st_size=4096, ...}, 0) = 0
newfstatat(AT_FDCWD, "/tmp/.dotnet/shm", {st_mode=S_IFDIR|0777, st_size=4096, ...}, 0) = 0
openat(AT_FDCWD, "/tmp/.dotnet/shm", O_RDONLY|O_CLOEXEC) = 83
flock(83, LOCK_EX)                      = -1 EACCES (Permission denied)
```

The file lock is taken on the `shm` directory to synchronize shared memory initialization and teardown between processes for the mutex. I'm not sure why `flock` would return `EACCES`, it doesn't seem to be documented as an error condition. By this point the permissions on the directory tree up to `shm` have successfully been verified to have the necessary access permissions.

There is a note that `flock` on a file descriptor over NFS may be based on `fcntl` and it may work differently (probably wouldn't work on a directory, though that error code is not documented for `fcntl` either). Seems like a bit of a stretch, but is `/tmp` a mount of an NFS share? A way to tell seems to be `cat /proc/mounts | grep nfs`.
